src/shared/passwdqc-dlopen.c

```diff
@@ -0,0 +1,144 @@
+
+        sym_passwdqc_params_reset(params);
+
+        r = sym_passwdqc_params_load(params, &load_reason, PASSWDQC_CONFIG);
+        if (r < 0) {
+                log_debug("Failed to load passwdqc configuration '%s', ignoring: %s",
+                          PASSWDQC_CONFIG, load_reason);
+        }
+
+        *ret = TAKE_PTR(params);
+        return 0;
+}
+
+#define N_SUGGESTIONS 6
+
+int suggest_passwords(void) {
+        _cleanup_(sym_passwdqc_params_freep) passwdqc_params_t *params = NULL;
+        _cleanup_strv_free_erase_ char **suggestions = NULL;
+        _cleanup_(erase_and_freep) char *joined = NULL;
+        size_t i;
+        int r;
+
+        r = pwqc_allocate_context(&params);
+        if (ERRNO_IS_NOT_SUPPORTED(r))
+                return 0;
+        if (r < 0)
+                return log_error_errno(r, "Failed to allocate libpasswdqc context: %m");
+
+        suggestions = new0(char*, N_SUGGESTIONS+1);
+        if (!suggestions)
+                return log_oom();
+
+        for (i = 0; i < N_SUGGESTIONS; i++) {
+                suggestions[i] = sym_passwdqc_random(params);
+                if (!suggestions[i])
+                        return log_error_errno(SYNTHETIC_ERRNO(EIO), "Failed to generate password, ignoring");
+        }
+
+        joined = strv_join(suggestions, " ");
+        if (!joined)
+                return log_oom();
+
+        log_info("Password suggestions: %s", joined);
+        return 1;
+}
+
+int check_password_quality(
+                const char *password,
+                const char *old,
+                const char *username,
+                char **ret_error) {
+        _cleanup_(sym_passwdqc_params_freep) passwdqc_params_t *params = NULL;
+        const char *check_reason;
+        int r;
+
+        assert(password);
+
+        r = pwqc_allocate_context(&params);
+        if (ERRNO_IS_NOT_SUPPORTED(r))
+                return 0;
+        if (r < 0)
+                return log_error_errno(r, "Failed to allocate libpasswdqc context: %m");
+
+        if (username) {
+                _cleanup_free_ char *name = NULL;
+                _cleanup_free_ char *empty = NULL;
+                name = strdup(username);
+                empty = strdup("");
+                if (!name || !empty)
+                        return -ENOMEM;
+
+                struct passwd pw = {
+                        .pw_name = name,
+                        .pw_gecos = empty,
+                        .pw_dir = empty
+                };
+                check_reason = sym_passwdqc_check(params, password, old, &pw);
+        } else
+                check_reason = sym_passwdqc_check(params, password, old, NULL);
+
+        if (check_reason != NULL) {
+                if (ret_error){
+                        _cleanup_free_ char *e = NULL;
+                        e = strdup(check_reason);
+                        if (!e)
+                                return -ENOMEM;
+                        *ret_error = TAKE_PTR(e);
```

Comment:
no need to do the `_cleanup_free_`/TAKE_PTR dance here given that there are no error paths where e is no-NULL